Improve error message in Kubelet CPU assignment logic

matte21 opened the pull request:
Include number of requested and available CPUs in the error message when the assignment of CPUs fails because there are less available CPUs than requested.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind documentation

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Include number of requested and available CPUs in the error message when the assignment of CPUs fails because there are less available CPUs than requested.

The new error message should speed up troubleshooting.

#### Special notes for your reviewer:

This change might be considered breaking, depending on whether the wording of error messages is guaranteed to be stable. i.e. if some client code examines the error by inspecting its message (e.g. via regexp), that client might break. The old wording was locked by unit tests, so I had to change some expected unit tests results.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
When the Kubelet fails to assign CPUs to a Pod because there less available CPUs than the Pod requests, the error message changed from
"not enough cpus available to satisfy request" to "not enough cpus available to satisfy request: <num_requested> requested, only <num_available> available".
```

matte21 committed:
Improve error message in cpu assignment logic

Include number of requested and available CPUs in the error message
when the assignment of CPUs fails because there are less available
CPUs than requested.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -453,7 +453,7 @@ func takeByTopologyNUMAPacked(topo *topology.CPUTopology, availableCPUs cpuset.C
 	}
 	if acc.isFailed() {
-		return cpuset.New(), fmt.Errorf("not enough cpus available to satisfy request")
+		return cpuset.New(), fmt.Errorf("not enough cpus available to satisfy request: %d requested, only %d available", numCPUs, availableCPUs.Size())
```
I'd go for a terser form: "requested %d, available %d" or even "requested=%d, available=%d". The current wording yeilds a nicer sentence, but IMO here practicality beats nice. Applies to everything below.

matte21 replied:
Done in [d4a5a08](https://github.com/kubernetes/kubernetes/pull/121059/commits/d4a5a085a84b3d2c3853b589787e25faec33e380)

ffromani reviewed (commented):
It's (still) unclear if we consider error messages part of the API contract. They probably are not, but the observability in this area is historically poor, so it's likely there's code/tooling (including tests) actually depending on the messages. Appending to the existing text is probably the best we can do.

There's an effort aiming to improve this aspect, stemming from the effort of GA'ing memory manager but pertaining to all resource managers (obv. incl. CPU manager). Let's see how this PR can fit in this picture.

swatisehgal commented:
/cc

matte21 commented:
The force push to [d4a5a08](https://github.com/kubernetes/kubernetes/pull/121059/commits/d4a5a085a84b3d2c3853b589787e25faec33e380) rebases on main and addresses the review comments.

matte21 commented:
/retest

matte21 commented:
@ffromani

ffromani commented:
/lgtm

I think it's a nice little improvement

matte21 commented:
/assign @[derekwaynecarr](https://github.com/derekwaynecarr)

matte21 commented:
/assign @derekwaynecarr

ffromani commented:
/cc @klueska @mrunalp

derekwaynecarr commented:
it is fine to update error messages.

/approve

saschagrunert reviewed (approved):
/lgtm
